src/cart/utils/memcheck-cart.supp

```diff
@@ -81,6 +81,82 @@
    fun:crt_hg_ctx_init
    ...
 }
+{
+   FI leak 6
+   Memcheck:Leak
+   match-leak-kinds: possible
+   ...
+   obj:/usr/lib64/librdmacm.so.1.1.22.1
+   ...
+}
+{
+   FI leak 7
+   Memcheck:Leak
+   match-leak-kinds: indirect
+   ...
+   fun:fi_ini
+   fun:fi_getinfo@@FABRIC_1.3
+   ...
+}
+{
+   FI leak 8
+   Memcheck:Leak
+   match-leak-kinds: possible
+   ...
+   fun:fi_ini
+   fun:fi_getinfo@@FABRIC_1.3
+   ...
+}
+{
+   FI leak 9
+   Memcheck:Leak
+   match-leak-kinds: reachable
+   ...
+   fun:fi_ini
+   fun:fi_getinfo@@FABRIC_1.3
+   ...
+}
+{
+   FI leak 10
+   Memcheck:Leak
+   match-leak-kinds: definite
+   ...
+   fun:fi_ini
+   fun:fi_getinfo@@FABRIC_1.3
+   ...
+}
+{
+   spdk leak 0
+   Memcheck:Leak
+   match-leak-kinds: reachable
+   fun:calloc
+   fun:rte_log_init
+   ...
+}
+{
+   spdk leak 1
+   Memcheck:Leak
+   match-leak-kinds: reachable
+   fun:calloc
+   fun:spdk_nvme_transport_register
+   ...
+}
+{
+   spdk leak 2
+   Memcheck:Leak
+   match-leak-kinds: reachable
+   fun:calloc
+   fun:spdk_rpc_register_method
+   ...
+}
+{
+   spdk leak 3
+   Memcheck:Leak
+   match-leak-kinds: reachable
+   fun:calloc
+   fun:spdk_rpc_register_alias_deprecated
```

Comment:
Are we using deprecated APIs?

Reply:
Yes, we are. The macro [SPDK_RPC_REGISTER](https://github.com/spdk/spdk/blob/b90630a465bf191b9bd3b0c2e4f2ba47b4044143/lib/rpc/rpc.c#L406) defines a constructor in [rpc.h](https://github.com/spdk/spdk/blob/b90630a465bf191b9bd3b0c2e4f2ba47b4044143/include/spdk/rpc.h#L130) that calls `spdk_rpc_register_alias_deprecated()`

Reply:
I'm not familiar with where we call these, and a quick "git grep" doesn't show anything.  Is there a ticket needed by us to update the API that we're using somewhere?

Reply:
These memory leaks are isolated to the SPDK libraries. Moreover, these functions are called using [constructors](https://www.geeksforgeeks.org/__attribute__constructor-__attribute__destructor-syntaxes-c/). 
I reproduced the memory leaks without using DAOS code. Here is the minimal example:

`#include <stdio.h>

int main(int argc, char *argv[]) {
  printf("Hello SPDK\n");

  return 0;
}`

Then, I built it and linked it to the spdk libraries

`$ gcc -g -L/home/jonathan/repos/daos/install/prereq/dev/spdk/lib -lspdk_rpc -lspdk_env_dpdk -lrte_bus_pci -o hello_spdk my_program.c`

Finally, I called valgrind

`$ valgrind --fair-sched=yes --leak-check=full --show-leak-kinds=all ./hello_spdk`

I got the same memory leaks:

==12888== 96 bytes in 2 blocks are still reachable in loss record 4 of 6
==12888==    at 0x4C2C089: calloc (vg_replace_malloc.c:762)
==12888==    by 0x4E38DEE: spdk_rpc_register_alias_deprecated (rpc.c:300)
==12888==    by 0x400F9B2: _dl_init (in /usr/lib64/ld-2.17.so)
==12888==    by 0x4001179: ??? (in /usr/lib64/ld-2.17.so)

==12888== 249 bytes in 29 blocks are still reachable in loss record 5 of 6
==12888==    at 0x4C29F73: malloc (vg_replace_malloc.c:309)
==12888==    by 0x54E4AF9: strdup (in /usr/lib64/libc-2.17.so)
==12888==    by 0x6477DE2: rte_log_init (in /home/jonathan/repos/daos/build/external/dev/spdk/dpdk/build/lib/librte_eal.so.20.0)
==12888==    by 0x400F9B2: _dl_init (in /usr/lib64/ld-2.17.so)
==12888==    by 0x4001179: ??? (in /usr/lib64/ld-2.17.so)

[full output here](https://github.com/daos-stack/daos/files/5343882/minimal_example.txt)

As we can see, these functions are called automatically. According to [Valgrind documentation](https://www.valgrind.org/docs/manual/faq.html), the `still reachable` status is not necessary a true memory leak. I am wondering if it is a bug?

Reply:
Ok, so it shoulds like this is something that SPDK needs to or is working on.  What I wanted to know was that there wasn't any action required on our side to update to latest interfaces.